File: src/lib/ordem-permissions.ts
import type { OrdemSection as DbOrdemSection } from '@prisma/client'
import type { LeaderRole } from '@/types/leader-role'
import type { CategorySpec, OrdemSection } from '@/types/ordem-item'
import { CATEGORY_MAP, isOrdemCategoryKey } from '@/types/ordem-item'

const GROUP_ROLES: LeaderRole[] = [
  'Chefe de Agrupamento',
  'Chefe de Agrupamento Adjunto',
  'Secretário de Agrupamento',
  'Tesoureiro de Agrupamento',
  'Assistente de Agrupamento',
]

const SECTION_ROLES: LeaderRole[] = [
  'Chefe de Unidade',
  'Chefe de Unidade Adjunto',
  'Instrutor',
]

export type ProfileForAuth = {
  role: 'ADMIN' | 'LEADER' | 'MEMBER'
  roles: string[]
  section: DbOrdemSection | OrdemSection | null
}

export function hasGroupRole(profile: ProfileForAuth): boolean {
  return profile.roles.some((r) => (GROUP_ROLES as string[]).includes(r))
}

export function hasSectionRole(profile: ProfileForAuth): boolean {
  return profile.roles.some((r) => (SECTION_ROLES as string[]).includes(r))
}

/**
 * Whether `profile` can create/update items for `category` (+ section).
 * Admins can do anything. Section items require the user to belong to that section
 * AND hold a section-level role. Group items require a group-level role.
 */
export function canManageItem(
  profile: ProfileForAuth,
  category: CategorySpec,
  section: OrdemSection | null
): boolean {
  if (profile.role === 'ADMIN') return true

  if (category.scope === 'GROUP') {
    return hasGroupRole(profile)
  }

  if (category.scope === 'BOTH') {
    // No section → group-level decision; section set → must belong to it
    if (section == null) return hasGroupRole(profile)
    if (hasSectionRole(profile) && profile.section === section) return true
    // Group leaders can also create activities for any section
    if (hasGroupRole(profile)) return true
    return false
  }

  // SECTION-scoped
  if (!section) return false
  if (!hasSectionRole(profile)) return false
  return profile.section === section
}

/**
 * Categories this profile can log. Admins get all; others get the subset
 * matching their roles.
 */
export function allowedCategoriesFor(profile: ProfileForAuth): CategorySpec[] {
  if (profile.role === 'ADMIN') return Object.values(CATEGORY_MAP)
  const hasGroup = hasGroupRole(profile)
  const hasSection = hasSectionRole(profile) && profile.section != null
  return Object.values(CATEGORY_MAP).filter((c) => {
    if (c.scope === 'GROUP') return hasGroup
    if (c.scope === 'SECTION') return hasSection
    return hasGroup || hasSection // BOTH
  })
}

export function resolveCategory(key: string) {
  if (!isOrdemCategoryKey(key)) return null
  return CATEGORY_MAP[key]
}

File: apps/api/app/core/ordem_permissions.py
"""OS item permissions. Port of `@qtscout/core/ordem-permissions`."""

from dataclasses import dataclass

from app.core.ordem_categories import CATEGORY_MAP, CategorySpec, is_ordem_category_key

GROUP_ROLES = (
    "Chefe de Agrupamento",
    "Chefe de Agrupamento Adjunto",
    "Secretário de Agrupamento",
    "Tesoureiro de Agrupamento",
    "Assistente de Agrupamento",
)

SECTION_ROLES = (
    "Chefe de Unidade",
    "Chefe de Unidade Adjunto",
    "Instrutor",
)


@dataclass
class ProfileForAuth:
    role: str  # ADMIN | LEADER | MEMBER
    roles: list[str]
    section: str | None


def has_group_role(profile: ProfileForAuth) -> bool:
    return any(r in GROUP_ROLES for r in profile.roles)


def has_section_role(profile: ProfileForAuth) -> bool:
    return any(r in SECTION_ROLES for r in profile.roles)


def can_manage_item(profile: ProfileForAuth, category: CategorySpec, section: str | None) -> bool:
    if profile.role == "ADMIN":
        return True

    if category.scope == "GROUP":
        return has_group_role(profile)

    if category.scope == "BOTH":
        if section is None:
            return has_group_role(profile)
        if has_section_role(profile) and profile.section == section:
            return True
        return has_group_role(profile)

    # SECTION-scoped
    if not section:
        return False
    if not has_section_role(profile):
        return False
    return profile.section == section


def allowed_categories_for(profile: ProfileForAuth) -> list[CategorySpec]:
    if profile.role == "ADMIN":
        return list(CATEGORY_MAP.values())
    group = has_group_role(profile)
    section = has_section_role(profile) and profile.section is not None
    result = []
    for c in CATEGORY_MAP.values():
        if c.scope == "GROUP":
            keep = group
        elif c.scope == "SECTION":
            keep = section
        else:
            keep = group or section
        if keep:
            result.append(c)
    return result


def resolve_category(key: str) -> CategorySpec | None:
    return CATEGORY_MAP[key] if is_ordem_category_key(key) else None
